### DISCUSSION TITLE
Return of Markdown Component in v2

### DISCUSSION THREAD

**Discussion by Psypher1** [withastro/roadmap#620] [Proposal]
# Summary

It would be nice if the markdown component that was present in v1 would be returned.

# Background & Motivation
When I started using Astro, there was a Markdown component - `@astrojs/markdown-component`. This made it simple to render markdown from external source on an Astro site. The removal of this component has stopped me from upgrading to v2 - lack of support.

I have several sites that rely on this component and I recently noticed it is no longer available. For me, this component was invaluable.

While people in Discord have mentioned that I use other things to achieve the same, that would now entail making sure it doesn't clash with Astro, where the component didn't.

I understand maintaining something like that takes time and effort. I would be willing to maintain it myself if the core team can't. I just can't find the repo where it was. 

If someone could give me access to it, I would gladly close this issue.

# Goals
- Straightforward rendering of external markdown data in Astro
- Access to the markdown component repository
- Support for the same type of implementation in v2

# Example

This is what I have/had which is convnient:
```js
---
import Markdown from "@astrojs/markdown-component";

export async function fetch getData(){
   const res = await fetch('...url')
   const data = res.json()   
}

const {body_markdown} = Astro.props

---

 <Markdown content={body_markdown} />

```

Thank you for the time

**Comment by bluwy** [comment]
It is no longer available from `"astro"`, but it's available in `"@astrojs/markdown-component"`. The package should still work today, and if not it might be a bug. The code is in the monorepo: https://github.com/withastro/astro/tree/main/packages/markdown/component

I also brought up internally that we may want to officially drop support for it in Astro v3, so having third-party support for it would be great! Even though so, `@astrojs/markdown-component` currently uses some internal APIs that I'd like to clean up, and hopefully it should still work after Astro v3, just without continued feature additions.

**Reply by shishkin** [reply]
What was the reason in the first place to deprecate and remove the component? There are valid use cases to load markdown manually and then use the same pipeline that Astro uses to render it instead of building the whole rendering in parallel as [advised in the docs](https://docs.astro.build/en/guides/markdown-content/#fetching-remote-markdown).

**Reply by bluwy** [reply]
I don't know the full context as it was already deprecated when I started maintaining Astro, but from what I've heard it's because it doesn't work in SSR well.

**Reply by shishkin** [reply]
That is strange because remark/rehype pipeline works well on node. Would you consider to bring the `Markdown` component back into core?

**Reply by bluwy** [reply]
Just did a check and remark/rehype plugins only work when used in string form, not function form. And we've been recommending using the function form more because it resolves better. Using the function-form won't work in either SSG and SSR if I understand correctly.

It's unlikely that it's coming back to core.

**Reply by bluwy** [reply]
String form looks like:

```js
{
  markdown: {
     remarkPlugins: ['remark-toc']
    }
}
```

Function form looks like:

```js
import remarkToc from 'remark-toc'

{
  markdown: {
     remarkPlugins: [remarkToc]
    }
}
```

**Reply by shishkin** [reply]
Hm.. I use rehype plugins in function form in Astro today. I will check remark and remark-toc. Though I doubt it's an issue inherent in remark since it's based on unified pipeline that have function-form plugins as first-class.

Anything else preventing Markdown component from returning to core?

**Reply by bluwy** [reply]
It's not only remark and remark-toc that's having issues, anything in the `markdown` config that is non-serializable would be an issue. Though further looking into it, I stand corrected that it's only an issue in SSR. SSG shares the same "realm" with the Astro config so no serialization is needed.

But with that said, it's not about blockers preventing it returning to core. It's that there's no more plans to "un-deprecate" it. I'm also planning to remove some Astro-specific hacks that's needed for `<Markdown>` to work, but I'll make it in a way that `@astrojs/markdown-component` will continue to work.

**Comment by bluwy** [comment]
After some internal discussion, we've decided to drop support for `@astrojs/markdown-component` in Astro v3. There are community packages like https://github.com/natemoo-re/astro-remote and https://www.npmjs.com/package/@astropub/md that can be used in place of this. `@astrojs/markdown-component` itself should still work in v2 if needed.

We won't be moving it back to core, but feel free to fork it and maintain your own version of it.

**Reply by shishkin** [reply]
Would you mind sharing any rationale for that decision? It's really confusing to me, considering that Markdown is itself a major feature of Astro and also a core dependency for Content Collections.

Looking at the available alternatives:
* [astro-remote](https://github.com/natemoo-re/astro-remote) doesn't use the same markdown pipeline as Astro and doesn't share the config and plugins, so its functionality won't be consistent;
* [@astropub/md](https://github.com/astro-community/md/) reuses Astro markdown config, plugins and directly delegates to `@astrojs/markdown-remark` for rendering. The component is trivial and mostly deals with sharing the config. It would be even smaller and simpler was it part of core and had access to internal APIs directly. And users won't need to bother installing the package and configuring the integration.

I would even understand it if you decided to remove all of Markdown features from core entirely. But having invested so much in Markdown code-wise and refusing to give users the simplest and most obvious feature is puzzling 😕
